### PR TITLE
WarmupLrScheduler and configs

### PR DESCRIPTION
**Pull request by strint**
支持Resnet50的功能补全
- lr warmup
- flow.config  -> flow.backends.xxx
  -  collective_boxing.nccl_fusion_all_reduce_use_buffer
- function_conf -> nn.Graph.config 之前已经支持，这里review下
- amp and loss scale conf

### CODE REVIEW

**Comment by strint** [comment]
##  lazy lr warmup
PyTorch没有官方实现，自己自由定制。所以这里是我们特有的。

### nn.Graph下使用
```
lr_warmup_constent = {"type": "constant", "steps": 7, "multiplier": 0.4}
graph.add_optimizer(sgd1, lr_sch=cosine_lr1, lr_warmup=lr_warmup_constant)
```

### 计算逻辑
####  constant (multiplier)
```
double ConstantWarmupLearningRate(const ConstantWarmupConf& conf, double lr, int64_t train_step) {
  CHECK_GE(conf.warmup_batches(), 0);
  CHECK_GT(conf.multiplier(), 0);  
  CHECK_LT(conf.multiplier(), 1);
  if (train_step < conf.warmup_batches()) {
    return lr * conf.multiplier();  // multiplier为常量，lr指数递减
  } else {
    return lr;
  }
}
```

#### linear (multiplier)
```
double LinearWarmupLearningRate(const LinearWarmupConf& conf, double lr, int64_t train_step) {
  CHECK_GE(conf.warmup_batches(), 0);
  CHECK_GE(conf.start_multiplier(), 0);
  CHECK_LT(conf.start_multiplier(), 1);
  double start_multiplier = conf.start_multiplier();
  double multiplier = 1.0;
  if (train_step < conf.warmup_batches()) {
    multiplier =
        start_multiplier + (1.0 - start_multiplier) * (train_step * 1.0 / conf.warmup_batches());
  }  // multiplier 从 start_multiplier开始线性增大，直到变成1，lr递减（减少程度慢于指数递减）
  return lr * multiplier;
}
```

### 和lr_sch的关系 + job调度逻辑
pass中插入 scheduler_op，用来修改lr，属于warmup控制的step，使用warmup控制；之后使用lr_scheduler控制。
```
void LearningRateScheduleKernel::ForwardDataContent(
    const KernelCtx& ctx, const std::function<Blob*(const std::string&)>& BnInOp2Blob) const {
  const LearningRateScheduleOpConf& conf = this->op_conf().learning_rate_schedule_conf();
  const int64_t train_step = *BnInOp2Blob("train_step")->dptr<int64_t>();
  float learning_rate = conf.learning_rate();
  if (TriggerWarmup(conf, learning_rate, train_step)) {
    learning_rate = GetWarmupLearningRate(conf.warmup_conf(), learning_rate, train_step);
  } else if (conf.has_learning_rate_decay()) {
    learning_rate = GetDecayedLearningRate(conf.learning_rate_decay(), learning_rate, train_step);
  }
  *BnInOp2Blob("out")->mut_dptr<float>() = learning_rate;

  if (Global<ResourceDesc, ForSession>::Get()->enable_debug_mode()) {
    (*log_stream_) << std::to_string(train_step) << ", " << std::to_string(learning_rate) << "\n";
    log_stream_->Flush();
  }
}
```
```

File: python/oneflow/test/graph/test_graph_optimizer.py [hunk 1]
```diff
@@ -122,16 +122,20 @@ def forward(self, x):
         cosine_lr0 = flow.optim.lr_scheduler.CosineAnnealingLR(
             sgd0, steps=10, alpha=0.01
         )
+        lr_warmup_constent = {"type": "constent", "steps": 7, "multiplier": 0.4}
         cosine_lr1 = flow.optim.lr_scheduler.CosineAnnealingLR(
             sgd1, steps=100, alpha=0.1
         )
+        lr_warmup_linear = {"type": "linear", "steps": 17, "start_multiplier": 0.6}
```
Comment (strint) [inline python/oneflow/test/graph/test_graph_optimizer.py:129]: 为减少类型、应对变化（warmup很灵活、torch没有提供官方支持、让用户自由定制），这里让用户直接传python原生的dict。debug也简单，可以直接打印dict

**Comment by strint** [comment]
backends控制接口
```
        flow.backends.nccl.boxing_fusion_enable(True)
        flow.backends.nccl.boxing_fusion_enable_mixed(True)
        flow.backends.nccl.boxing_fusion_all_reduce(True)
        flow.backends.nccl.boxing_fusion_all_reduce_use_buffer(True)
        flow.backends.nccl.boxing_fusion_reduce(True)
        flow.backends.nccl.boxing_fusion_broadcast(True)
        flow.backends.nccl.boxing_fusion_all_gather(True)
        flow.backends.nccl.boxing_fusion_reduce_scatter(True)
        flow.backends.nccl.boxing_fusion_threshold_mb(1024)
        flow.backends.nccl.boxing_fusion_max_ops_num(5)
        flow.backends.nccl.streams_num(4)
        flow.backends.nccl.reuse_compute_stream(True)
        flow.backends.nccl.enable_all_to_all(True)

        flow.backends.cuda.reserved_mem_mbyte(1000)

        flow.backends.debug.enable(True)
        flow.backends.lib.load("")
```

**Comment by liujuncheng** [comment]
> backends控制接口【分类控制，待讨论】
> 
> ```
>         flow.backends.nccl.boxing_fusion_enable(True)
>         flow.backends.nccl.boxing_fusion_enable_mixed(True)
>         flow.backends.nccl.boxing_fusion_all_reduce(True)
>         flow.backends.nccl.boxing_fusion_all_reduce_use_buffer(True)
>         flow.backends.nccl.boxing_fusion_reduce(True)
>         flow.backends.nccl.boxing_fusion_broadcast(True)
>         flow.backends.nccl.boxing_fusion_all_gather(True)
>         flow.backends.nccl.boxing_fusion_reduce_scatter(True)
>         flow.backends.nccl.boxing_fusion_threshold_mb(1024)
>         flow.backends.nccl.boxing_fusion_max_ops_num(5)
>         flow.backends.nccl.streams_num(4)
>         flow.backends.nccl.reuse_compute_stream(True)
>         flow.backends.nccl.enable_all_to_all(True)
> 

这里感觉nccl和boxing反了
应该是 
```
 flow.backends.boxing.nccl_fusion_enable(True)
``` 

>        flow.backends.cuda.reserved_mem_mbyte(1000)
这个目前应该没有实际用途，可以先不导出到新接口

>        flow.backends.debug.enable(True)
这个也可以不导出，就环境变量控制挺好的

File: python/oneflow/nn/graph.py [hunk 2]
```diff
@@ -96,6 +106,9 @@ def add_optimizer(
                 lr_sch._optimizer is optim
             ), "lr_scheduler's optimizer must be the same optimizer in add_optimizer."
             opt_dict["lr_sch"] = lr_sch
+        if lr_warmup is not None:
+            assert isinstance(lr_warmup, dict)
```
Comment (ShawnXuan) [inline python/oneflow/nn/graph.py:110]: warmup的过程也是lr schedule过程中的一步。lr schedule的过程分两步：warmup和decay，warmup的配置也可以在LrScheduler构造时候传入。

退一步来说，即使不放到LrScheduler里面，也不建议传dict，可以再定义一个LrWarmUp

**Comment by strint** [comment]
graph.config
```
# # 优化参数
# enable_cudnn_fused_normalization_add_relu
# enable_fuse_add_to_output
# enable_fuse_cast_scale
# do_parallel_cast_before_widening_type_cast
# optimizer_placement_optimization_mode
# optimizer_placement_optimization_threshold
# enable_non_distributed_optimizer
# prune_parallel_cast_ops
# prune_cast_to_static_shape_ops
# prune_amp_white_identity_ops
# enable_auto_mixed_precision
# enable_fuse_model_update_ops
# enable_gradients_stats_aggregation
# train.num_gradient_accumulation_steps
```

File: python/oneflow/nn/graph.py (hunk 2 above)
Comment (strint) [reply python/oneflow/nn/graph.py:110]: ok，这里会采用WarmupLrScheduler的做法

**Comment by leaves-zwx** [comment]
根据咨询 juncheng 和 shenghang 和我们下午讨论的结果，nccl 相关的配置的接口建议如下：

```
flow.boxing.nccl.set_stream_num(int)
flow.boxing.nccl.set_fusion_threshold_mb(int)
flow.boxing.nccl.set_fusion_max_ops(int)
flow.boxing.nccl.can_fuse_all_reduce(bool)
flow.boxing.nccl.can_fuse_reduce_scatter(bool)
flow.boxing.nccl.can_fuse_all_gather(bool)
flow.boxing.nccl.can_fuse_reduce(bool)
flow.boxing.nccl.can_fuse_broadcast(bool)
flow.boxing.nccl.can_fuse_mixed_ops(bool)
flow.boxing.nccl.use_buffer_to_fuse_all_reduce(bool)
flow.boxing.nccl.enable_all_to_all(bool)
```

File: python/oneflow/nn/optimizer/lr_scheduler.py [hunk 3]
```diff
@@ -76,3 +78,31 @@ def step(self):
             group["lr"] = self.last_lr[i]
             if self.verbose:
                 self.print_lr(i, self.last_lr[i])
+
+
+class WarmupLrScheduler(LrScheduler):
```
Comment (strint) [inline python/oneflow/nn/optimizer/lr_scheduler.py:83]: WarmupLrScheduler

File: python/oneflow/test/graph/test_graph_optimizer.py [hunk 4]
```diff
@@ -122,16 +122,22 @@ def forward(self, x):
...
         )
+        constant_warmup_cosine_lr0 = flow.optim.lr_scheduler.ConstantWarmupLR(
+            cosine_lr0, steps=7, multiplier=0.4
+        )
         cosine_lr1 = flow.optim.lr_scheduler.CosineAnnealingLR(
             sgd1, steps=100, alpha=0.1
         )
+        linear_warmup_cosine_lr1 = flow.optim.lr_scheduler.LinearWarmupLR(
+            cosine_lr1, steps=17, start_multiplier=0.6
+        )
```
Comment (strint) [inline python/oneflow/test/graph/test_graph_optimizer.py:133]: WarmupLrScheduler的使用方式

File: python/oneflow/nn/optimizer/lr_scheduler.py [hunk 5]
```diff
@@ -76,3 +78,31 @@ def step(self):
...
+    def __init__(
+        self, lrsch_or_optimizer, last_step=-1, verbose=False,
+    ):
+        self._lr_sch = None
+        isinstance(lrsch_or_optimizer, (LrScheduler, Optimizer))
+        if isinstance(lrsch_or_optimizer, LrScheduler):
+            super().__init__(lrsch_or_optimizer._optimizer, last_step, verbose)
+            self._lr_sch = lrsch_or_optimizer
+        else:
+            super().__init__(lrsch_or_optimizer, last_step, verbose)
```
Comment (strint) [inline python/oneflow/nn/optimizer/lr_scheduler.py:93]: 支持接受一个Optimizer 或者 LrScheduler

**Comment by strint** [comment]
> 根据咨询 juncheng 和 shenghang 和我们下午讨论的结果，nccl 相关的配置的接口建议如下：
> 
> ```
> flow.boxing.nccl.set_stream_num(int)
> flow.boxing.nccl.set_fusion_threshold_mb(int)
> flow.boxing.nccl.set_fusion_max_ops(int)
> flow.boxing.nccl.can_fuse_all_reduce(bool)
> flow.boxing.nccl.can_fuse_reduce_scatter(bool)
> flow.boxing.nccl.can_fuse_all_gather(bool)
> flow.boxing.nccl.can_fuse_reduce(bool)
> flow.boxing.nccl.can_fuse_broadcast(bool)
> flow.boxing.nccl.can_fuse_mixed_ops(bool)
> flow.boxing.nccl.use_buffer_to_fuse_all_reduce(bool)
> flow.boxing.nccl.enable_all_to_all(bool)
> ```

can -> allow ?

参考 https://pytorch.org/docs/stable/backends.html?highlight=allow#torch.backends.cudnn.allow_tf32

File: python/oneflow/test/graph/test_sys_opt_conf.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        flow.backends.debug.enable(True)
+        flow.backends.lib.load("")
+
+        class CustomGraphSysConf(flow.nn.Graph):
+            def __init__(self):
+                super().__init__()
+                self.m = CustomModule()
+                self.config.enable_amp(True)
+                loss_scale = flow.nn.graph.amp.DynamicLossScalePolicy(3000, 1000, 3.0)
+                self.config.amp_add_loss_scale_policy(loss_scale)
```
Comment (strint) [inline python/oneflow/test/graph/test_sys_opt_conf.py:82]: amp and loss scale

**Comment by strint** [comment]
这两天 @chengtbf @leaves-zwx 我们一起refine graph要暴露的接口，取舍时很抓头，有了一点头绪，总结这里的取舍思路。

#### 创造概念or命名时

1. 符合原有流行软件的命名习惯，没有再自创；
2. 要简单，简单才能大部分人一起玩耍、才能流行，“强大但复杂的东西”会被 “强大且简单的东西”取代，代码、概念都有这样的现象；
3. 总是要符合英文/中文语言习惯和语法，读起来自然，代码和文档的可读性不高，后面还是会被重构掉；
4. 创建新概念时，有分类，有namescope；
5. 非常常用时，可以减少namescope的层级；

**Review by wyg1997 — APPROVED** [review]

**Comment by chengtbf** [comment]
nn.Graph.config.xx:

```python
enable_fuse_model_update_ops
enable_fuse_add_to_output
enable_fuse_cast_scale
enable_cudnn_fused_normalization_add_relu
num_gradient_accumulation_steps
enable_auto_mixed_precision
enable_quantization_aware_training
```

**Comment by strint** [comment]
flow.backends.cudnn.enable_fused_normalization_add_relu

这个这次发布强依赖不？

它从graph.config改到backends，需要动下面的数据结构。

**Comment by strint** [comment]
这个分支保留作为resnet50的测试分支。等另外一个正式分支合并了就废弃。
